### PR TITLE
Stripe Payouts

### LINKED ISSUE #86: stripe payout/payment refactoring
refactor payment workflow so that fees are added to connected accounts immediately, and then we get our cut, which is then taxed by stripe. work the payout workflow so that renters get their money transferred on a regular basis.

### PR DESCRIPTION
**Pull request by nishantbalaji**
Closes #86

### CODE REVIEW

**Comment by nishantbalaji** [comment]
Test this concurrently with the PR in the Frontend directory

**Comment by nishantbalaji** [comment]
Changes:

- route to connect to stripe for payouts
- route to access stripe dashboard

File: routes/payment.js [hunk 1]
```diff
@@ -4,17 +4,18 @@ const router = express.Router();
...
+// const stripePublicKey = process.env.STRIPE_PUBLIC_KEY
 
 const stripe = require('stripe')(stripeSecretKey);
 
 // Create stripe checkout session
 router.post('/create-session', async (req, res) => {
   try {
     const { dates, days, listingId, reservationId } = req.body
-
+    console.log(listingId);
```
Comment (aidenszeto) [inline routes/payment.js:18]: Remove console log

File: routes/payment.js [hunk 2]
```diff
@@ -42,31 +56,47 @@ router.post('/create-session', async (req, res) => {
...
+    const stripeId = user.stripeId;
+    const guestFee = reservationDetails.guestFee;
+    const listingPrice = listingDetails.price;
+
+
+    // Total price of the listing including guest fee. Host fee + our profit is taken from this. 
+    const listingTotal = (listingPrice * days * 100) + (guestFee * 100);
+
+    // Price we make - Host Fee (1%) + Guest Fee (10%)
+    const applicationFee = (listingPrice * days * 100) * 0.01 + (guestFee * 100)
```
Comment (aidenszeto) [inline routes/payment.js:75]: Line 72 and 75 - we should make these constants into global variables so we don't have any magic numbers.

Correct me if I'm wrong:
 - 0.01 represents HOST_FEE
 - 100 represents CENTS_TO_DOLLARS

File: routes/payout.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+const express = require('express');
+const { rest } = require('lodash');
```
Comment (aidenszeto) [inline routes/payout.js:2]: Doesn't look like we're using this import

File: routes/payout.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      console.log(e)
+      res.status(500).json({
+        'error': 'Stripe accountLink failed'
+      })
+    } 
+}
+
+);
+
+router.post('/express', async(req, res) => {
```
Comment (aidenszeto) [inline routes/payout.js:48]: What does this route do? Could we add some comments to document both the routes in `payout.js`

**Review by aidenszeto — CHANGES_REQUESTED** [review]

File: routes/payment.js [hunk 5]
```diff
@@ -66,13 +64,13 @@ router.post('/create-session', async (req, res) => {
     const stripeId = user.stripeId;
     const guestFee = reservationDetails.guestFee;
     const listingPrice = listingDetails.price;
-
+    const centsToDollars = 100;
```
Comment (aidenszeto) [inline routes/payment.js:67]: We should keep constants like these in all caps (`CENTS_TO_DOLLARS`)

File: routes/payment.js [hunk 6]
```diff
@@ -66,13 +64,13 @@ router.post('/create-session', async (req, res) => {
...
-
+    const centsToDollars = 100; 
 
     // Total price of the listing including guest fee. Host fee + our profit is taken from this. 
-    const listingTotal = (listingPrice * days * 100) + (guestFee * 100);
+    const listingTotal = (listingPrice * days * centsToDollars) + (guestFee * centsToDollars);
 
     // Price we make - Host Fee (1%) + Guest Fee (10%)
-    const applicationFee = (listingPrice * days * 100) * 0.01 + (guestFee * 100)
+    const applicationFee = (listingPrice * days * centsToDollars) * 0.01 + (guestFee * centsToDollars)
```
Comment (aidenszeto) [inline routes/payment.js]: The 0.01 should also be changed to a constant. We should move all global variables to the top of the file as well

File: routes/payment.js (hunk 5 above)
Comment (aidenszeto) [reply routes/payment.js:67]: Could we also move these constants to the top of the file since they're global

File: routes/payout.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      const accountLink = await stripe.accountLinks.create({
+      account: account.id,
+      refresh_url: `${baseURL}/myAccount`,
+      return_url: `${baseURL}/myAccount`,
+      type: 'account_onboarding',
+      });
+
+      const redirectURL = accountLink['url'];
+
+      const user = await User.findOne({ email });
```
Comment (aidenszeto) [inline routes/payout.js:30]: What happens if the email provided is not associated with a valid user? We need a validation check here that returns a 400 on invalid email to ensure that the next line (line 31) does not break.

Comment (aidenszeto) [reply routes/payout.js:30]: Another option is to just use `findOneAndUpdate` to bypass the potential issues on line 31. If we choose to use `findOneAndUpdate`, we should still add a validation check to see if the email is valid or not though.

**Review by aidenszeto — CHANGES_REQUESTED** [review]

**Review by aidenszeto — APPROVED** [review]

**Pull request merged**